launcher/minecraft/PackProfile.cpp

```diff
@@ -954,8 +988,58 @@ bool PackProfile::setComponentVersion(const QString& uid, const QString& version
         ComponentPtr component = *iter;
         // set existing
         if (component->revert()) {
+            // remove linked components to let them re-resolve their versions
+            if (important) {
+                component->waitLoadMeta();
+                auto linked = collectTreeLinked(uid);
+                for (auto comp : linked) {
+                    if (comp->isCustom()) {
+                        continue;
+                    }
+                    if (modloaderMapping.contains(comp->getID())) {
+                        qCDebug(instanceProfileC)
+                            << d->m_instance->name() << "|" << comp->getID() << "is a mod loader, updating it to a compatible version";
+
+                        auto versionList = APPLICATION->metadataIndex()->get(comp->getID());
+                        versionList->waitToLoad();
+                        if (versionList) {
+                            auto recommended = versionList->getRecommendedForParent(uid, version);
+                            if (recommended) {
+                                qCDebug(instanceProfileC) << d->m_instance->name() << "|"
+                                                          << "setting updated loader to recommended version: " << comp->getID() << " = "
+                                                          << recommended->version();
+                                comp->setVersion(recommended->version());
+                            } else {
+                                auto latest = versionList->getLatestForParent(uid, version);
+                                if (latest) {
+                                    qCDebug(instanceProfileC) << d->m_instance->name() << "|"
+                                                              << "setting updated loader to latest compatible version: " << comp->getID()
+                                                              << " = " << latest->version();
+                                    comp->setVersion(latest->version());
+                                } else {
+                                    qCDebug(instanceProfileC) << d->m_instance->name() << "|"
+                                                              << "no compatible version for" << comp->getID() << "removing";
+                                    remove(comp->getID());
```

Comment:
I would argue that in this case, we should either display a message to the user to signal an error or at least not remove it(unless maybe is disabled) 
As the user may be offline removing the modloader would cause trouble.